Toponaming: Improve toponaming support for the python API.

drwho495 opened the pull request:
fixes: https://github.com/freeCAD/FreeCAD/issues/18671

This PR implements a few changes found in realthunder's fork to improve toponaming support for addons (like sheetmetal) so that users have a more stable experience when designing parts.

Please try out the changes with [this](https://github.com/drwho495/FreeCAD/releases/tag/weekly-2025.10.13) development build.

drwho495 committed:
fix getElement, Face, Compound, Edge, Wire,  ancestorsOfType, TopoShape, and more

drwho495 committed:
fix getElement, Face, Compound, Edge, Wire,  ancestorsOfType, TopoShape, and more

drwho495 committed:
Merge branch 'FreeCAD:main' into fix-addon-tpn

drwho495 committed:
add missing code to the mapper and fix bug from last commit

drwho495 committed:
add missing code to the mapper and fix bug from last commit

drwho495 committed:
add flush call to _getTopoShape in the cache

the cache is black magic and i don't know what i'm doing with it

drwho495 committed:
add flush call to _getTopoShape in the cache

the cache is black magic and i don't know what i'm doing with it

drwho495 committed:
Merge branch 'main' into fix-addon-tpn

drwho495 committed:
only flush if the parent's cached element map isn't null

drwho495 committed:
only flush if the parent's cached element map isn't null

drwho495 committed:
add ensureElementMap call

drwho495 committed:
add ensureElementMap call

drwho495 committed:
Merge branch 'FreeCAD:main' into fix-addon-tpn

drwho495 committed:
fix windows error

don't edit the ts variable, it causes an error (access violation?)

drwho495 committed:
fix windows error

don't edit the ts variable, it causes an error (access violation?)

drwho495 committed:
Merge branch 'fix-addon-tpn' of https://github.com/drwho495/FreeCAD into fix-addon-tpn

drwho495 committed:
Merge branch 'FreeCAD:main' into fix-addon-tpn

chennes commented:
@drwho495 do you need help resolving those test failures?

Connor9220 commented:
> > @drwho495 do you need help resolving those test failures?
> 
> I do. It looks like they are related to CAM offsetting a wire. I'm not really sure how or if toponaming changes causes this.

I've looked at this.  Looks like for what ever reason, the offset code is flipping a wire and thus the assertions are not correct anymore.  Why is offset flipping the wire?  I'm not sure, this would indicate that the base geometry is different now then when the test was written. Is topo is changing it somehow?  I looked at the offset, it looks fine visually.  We can update the test to accommodate this, but is this something we need to be worried about?

drwho495 committed:
Merge branch 'FreeCAD:main' into fix-addon-tpn

maxwxyz commented:
should this be in 1.1 or for 1.2?

drwho495 committed:
modify CAM tests

drwho495 committed:
modify CAM tests

drwho495 commented:
> should this be in 1.1 or for 1.2?

i would like this to be implemented for 1.1, i have been asked by multiple sheet metal users about this, it's one of the biggest toponaming pain points for many

drwho495 commented:
@Connor9220 i went ahead and modified the tests that failed. i think the offset flipping could be related to freecad now returning a direct copy of a TopoShape from the c++ side to the python side, before it would return a TopoDS_Shape which was converted to a TopoShape. (maybe tangents of edges are flipped during the conversion from the TopoDS_Shape to the TopoShape?)

drwho495 committed:
Merge branch 'FreeCAD:main' into fix-addon-tpn

drwho495 committed:
fix wire generation

drwho495 committed:
fix wire generation

drwho495 committed:
Merge branch 'FreeCAD:main' into fix-addon-tpn

drwho495 committed:
fix face generation code for BIM test

drwho495 committed:
Merge branch 'fix-addon-tpn' of https://github.com/drwho495/FreeCAD into fix-addon-tpn

drwho495 committed:
Merge branch 'main' into fix-addon-tpn

adrianinsaval commented:
> @Connor9220 i went ahead and modified the tests that failed. i think the offset flipping could be related to freecad now returning a direct copy of a TopoShape from the c++ side to the python side, before it would return a TopoDS_Shape which was converted to a TopoShape. (maybe tangents of edges are flipped during the conversion from the TopoDS_Shape to the TopoShape?)

#23498 related?

drwho495 commented:
> @drwho495 this one makes us a little nervous at this late date, so I think we're going to move it to 1.2, with the possibility of backporting it to 1.1.x if when merged it doesn't break anything. Can you provide a couple of example files that we can turn into tests?

@shaise would have some. I'm going to retrigger a weekly build of that branch so that it can be retested with the latest changes.

kadet1090 commented:
Unfortunately in case of that PR formatting changes are not the only conflicts so I couldn't update it automatically. Instruction on how to best deal with such massive automatic changes to the codebase is here: https://github.com/FreeCAD/FreeCAD/discussions/25217. 

While it would be awesome for you to do all the steps necessary, if you resolve only conflicts before changes that is fine too - I can take care of formatting changes. In case of any troubles, feel free to ping me here (or on any other platform) and I'll try to help.

drwho495 commented:
> ping for the conflicts (resolution above)

I'll look into this soon, I'm a bit busy right now unfortunately.

drwho495 committed:
Merge branch 'main' into fix-addon-tpn

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

drwho495 committed:
fix merge bugs

drwho495 committed:
Merge branch 'fix-addon-tpn' of https://github.com/drwho495/FreeCAD into fix-addon-tpn

Reqrefusion commented:
CL seems to have failed.

chennes committed:
Part: Fix curly braces in TopoShapeFacePyImp.cpp

chennes committed:
Part/TopoShapeFace: Don't add first wire twice

Reqrefusion commented:
@drwho495, were you able to take a look?

sliptonic commented:
It looks like the offsetWire method and associated tests have changed significantly since the realthunder code was written.  This is an extremely important method for CAM.

tarman3 commented:
Applied changes from this PR without changes in `TestPathOpUtil.py`
`TestPathOpUtil` still pass without errors

Probably no reason to touch `TestPathOpUtil`

drwho495 commented:
> I tested this PR locally and found some issues.

Not sure what's going on with the ancestors but the compound bug should be an easy fix.

maxwxyz commented:
@drwho495 friendly reminder for the conflict

drwho495 commented:
> @drwho495 friendly reminder for the conflict

Thanks. I should be able to look at it tomorrow.

drwho495 committed:
Merge branch 'main' into fix-addon-tpn

drwho495 committed:
Fix compound

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

drwho495 committed:
Prevent duplicates in ::ancestorsOfType

drwho495 committed:
Merge branch 'fix-addon-tpn' of https://github.com/drwho495/FreeCAD into fix-addon-tpn

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

drwho495 commented:
I've fixed the issues with ancestors and compounds. I also removed the change to the CAM test. The tests pass, so feel free to test this with sheetmetal files.

hyarion commented:
@Connor9220 could you look at it please?

Reqrefusion commented:
@hyarion, have you had a chance to check it out?

hyarion commented:
It is a bit difficult to follow the reason behind some of the changes and it isn't obvious from the commits either. Right now to review the changes we don't really have much context from you and instead we need to figure out ourselves why you did the changes.

I'm not sure if I'm asking too much, but it would be really nice if you could restructure the commits and explain in the commit messages what the change does, and/or write tests that shows the new behavior. Would that be possible?

Reqrefusion commented:
> It is a bit difficult to follow the reason behind some of the changes and it isn't obvious from the commits either. Right now to review the changes we don't really have much context from you and instead we need to figure out ourselves why you did the changes.
> 
> I'm not sure if I'm asking too much, but it would be really nice if you could restructure the commits and explain in the commit messages what the change does, and/or write tests that shows the new behavior. Would that be possible?

@drwho495 ?

Reqrefusion commented:
Will you share this sacred secret of toponomaming with us mortals, @drwho495?

drwho495 commented:
> Will you share this sacred secret of toponomaming with us mortals, @drwho495?

I've been working on some basic documentation, but I haven't found the time to finish it. Quite frankly I don't remember all of the code in this PR (especially since it's almost a year old). Most of the changes consist of using the proper make element methods from TopoShapeExpansion and using TopoShapes.

Reqrefusion commented:
I agree, this is a change that has been open for a long time and actually fixes important things. Doing the documentation work later isn't such an unfamiliar solution.

pieterhijma commented:
@drwho495, we've been looking into this PR and we will merge this, but it would be really good if you could provide the documentation nonetheless or some regression tests. We are all eager to understand more about toponaming!

maxwxyz commented:
@PaddleStroke will this be a benefit for the Assembly WB?

FlachyJoe commented on src/Mod/Part/App/TopoShapeCache.cpp:
```diff
@@ -78,13 +83,19 @@ TopoShape TopoShapeCache::Ancestry::_getTopoShape(const TopoShape& parent, int i
+    // the subelement doesn't have an element map, lets find it by mapping the parent shape onto the
+    // subelement.
+    if (res.elementMap(false) == 0 && res._parentCache->cachedElementMap) {
+        res.flushElementMap();
```
This line causes a crash because `topoShapes` is cleared so the  line 44
```   TopoShape& foundTS = topoShapes[index - 1];```
raised an assert the next time it's called.

```
1  Part::TopoShapeCache::Ancestry::_getTopoShape TopoShapeCache.cpp     97   0x7fffa975cd91 
2  Part::TopoShapeCache::Ancestry::getTopoShapes TopoShapeCache.cpp     125  0x7fffa975d0a0 
3  Part::TopoShape::getSubTopoShapes             TopoShapeExpansion.cpp 1154 0x7fffa9764517 
4  Part::WireJoiner::addShape                    WireJoiner.cpp         3098 0x7fffa97ff0aa 
5  Sketcher::SketchObject::buildInternals        SketchObject.cpp       414  0x7fff3e1e24d7 
6  Sketcher::SketchObject::buildShape            SketchObject.cpp       358  0x7fff3e1e15c2 
7  Sketcher::SketchObject::execute               SketchObject.cpp       244  0x7fff3e1e0621 
8  App::DocumentObject::recompute                DocumentObject.cpp     166  0x7ffff3871a03 
9  Part::Feature::recompute                      PartFeature.cpp        112  0x7fffa93d10d5 
10 App::Document::_recomputeFeature              Document.cpp           3223 0x7ffff37db977
```

FlachyJoe replied:
It seems to only occurs at sketch internal faces creation.

FlachyJoe replied:
Will be fixed by #30467
